ESPS/general/src/featohtk/featohtk.1
.\" Copyright (c) 1998 Entropic Research Laboratory, Inc. All rights reserved.
.\" @(#)featohtk.1	1.1 3/17/98 ERL
.ds ]W (c) 1998 Entropic Research Laboratory, Inc.
.if n .ds - ---
.if t .ds - \(em\h'-0.5m'\(em
.TH  FEATOHTK 1\-ESPS 3/17/98
.SH NAME

featohtk \- HTK input filter for ESPS FEA files

.SH SYNOPSIS
.B
featohtk
[
.BI \-f " field"\c
.RI [ range ]
] [
.BI \-k " parmKind"
] [
.BI \-r " range"
] [
.BI \-x " debug_level"
] [
.BI \-E " field"\c
.RI [ index ]
] [
.B \-L
] [
.BI \-O " field"\c
.RI [ index ]
] [
.BI \-P " param_file"
]
.I "input output"
.SH DESCRIPTION
.PP
.I featohtk
reads a FEA file
.I input,
extracts the contents of specified fields or portions of fields,
formats the extracted data as an HTK waveform or parameter file
(see ref. [1]),
and writes the result as
.I output.
The function is designed to be used as an input filter to permit
convenient reading by HTK of data written by ESPS programs or
.I waves+.
Output is in HTK default most-significant-byte-first order, regardless
of the architecture of the machine on which the program is run.
.PP
If
.I input
is ``\-'', standard input is used for the input file.
If
.I output
is ``\-'', standard output is used for the output file.
The input and output should not be the same file;
however, the program can be run as a filter
by specifying ``\-'' for both input and output.
.PP
The output consists of a standard HTK header
(see section 5.7.1 [``HTK Format Parameter Files''] of ref. [1])
followed by a sequence of waveform samples
or parameter vector samples.
.PP
The header contains items that indicate
the number of samples in the file,
the sample period,
the number of bytes per sample,
and the parameter kind.
The sample period is given in units of 100 nanoseconds
and is computed from the ESPS generic header item
.I record_freq
in the input header.
The sample kind is a code indicating whether the samples
are waveform samples (WAVEFORM) or contain vectors of
linear-prediction filter coefficients (LPC),
mel-frequency cepstral coefficients (MFCC), etc.
.PP
The sample kind is determined by command-line options
or ESPS parameter-file entries; see
.B \-k
under OPTIONS for the full list of supported basic parameter kinds.
For parameter kinds other than WAVEFORM, IREFC, and DISCRETE,
the basic parameter vector can be augmented with a log-energy term;
see option
.BR \-E .
For parameter kind MFCC, a ``zeroth-order cepstral parameter''
may appear in addition to, or in place of, the log energy;
see option
.BR \-O .
.PP
Each input record supplies the contents for one output waveform sample
or parameter sample vector.
The basic output coefficients (waveform samples or parameter vectors)
are taken from a specified field;
see option
.BR \-f .
The entire field or a specified subset may be used.
The log-energy term, if any, may be a specified element of the same
field or another field;
see option
.BR \-E .
Likewise the zeroth-order cepstral parameter, if included,
may be a specified element of the same field or another;
see option
.BR \-O .
.PP
In the output vector, the basic coefficients (specified with
.BR \-f )
come first, followed by the zeroth-order cepstral term, if any,
followed finally by the log-energy term, if any.
.PP
The data type of the output vectors is
.I short
for parameter kinds WAVEFORM, IREFC, and DISCRETE, and
.I float
for the other parameter kinds.
The data types of the input fields
need not match the output data type; 
conversions are performed as necessary.
.SH OPTIONS
.PP
The following options are supported.
Default values are shown in parentheses.
.TP
.BI \-f " field" "\h'|1.4i'\fR(no default)\fP"
.TP
\fB\-f "\fIfield\fB\^[\^\fIgrange\fB\^]\^"\fR
The name of the field containing the waveform sample values
or parameter values to be included in the output.
.IP
The second form, with brackets,
indicates a subset of the field elements
that are to be included in the output.
For example, ``\-f "auto[0,2:5,8]"'' indicates that the output vector
contains elements number 0, 2 through 5, and 8
of the input field named
.I auto.
The reason for the double quotes is simply to prevent the shell
from giving the usual special interpretation to the brackets.
The brackets contain a general range specification
of the form handled by the function
.IR grange_switch (3\-ESPS).
See the function manual page for the detailed syntax.
.IP
The first form of the option argument, without the bracketed range,
denotes the entire field.
.TP
.BI \-k " parmKind" "\h'|1.4i'\fR(no default)"
A string indicating the basic kind of data
to be indicated in the output HTK header\*-e.g. WAVEFORM
for sampled waveform data
or LPREFC for linear-prediction reflection coefficients.
The allowed values are:
.sp 0.5
.nf
.ta 0.25i 1.50i 2.75i 4.0i
	WAVEFORM	LPC	LPREFC	LPCEPSTRA
	IREFC	MFCC	FBANK	MELSPEC
	USER	DISCRETE
.DT
.fi
.sp 0.5
For the meaning of each,
see section 5.7.1 (``HTK Format Parameter Files'') of ref. [1].
(Note that LPDELCEP, though listed in that section, is not supported.)
.TP
.BI \-r " start" : "last" "\h'|1.4i'\fR(1:[last in file])\fP"
.TP
.BI \-r " start" :+ "incr"
.TP
.BI \-r " start"
The range of records to be processed.
In the first form, a pair of unsigned integers
gives the record numbers
of the first and last records to be processed
(counting from 1 for the first record of the file).
Either
.I start
or
.I last
may be omitted; then the default value is used.
If
.IR last " = " start " + " incr,
the second form (with the plus sign) specifies the same range as the first.
The third form (omitting the colon) specifies a single record.
.TP
.BI \-x " debug_level" "\h'|1.4i'\fR(0)\fP"
A positive value of
.I debug_level
calls for debugging output,
which is printed on the standard error output.
Larger values result in more output.
For the default value of 0, no messages are printed.
.TP
.BI \-E " field"
.TP
\fB\-E "\fIfield\fB\^[\^\fIindex\fB\^]\^"\fR
A field element containing an energy or log energy term
to be included in an output parameter file in addition to the
basic parameters in the field specified with
.BR \-f .
The resulting output file has a parameter kind
denoted by appending the qualifier ``_E''
to the basic type indicated with
.BR \-k.
For example, if
.BR \-k " LPREFC"
was specified, the output file would have the format indicated by
``LPREFC_E''.
See section 5.5 (``Energy Measures'') of ref. [1].
In the second form, the
.I index
in brackets is an integer that indicates the field element to be used
(counting from 0 for the first element).
(The reason for the double quotes is simply to prevent the shell
from giving the usual special interpretation to the brackets.)
The first form, without brackets, may be used
when the field has only one element.
.TP
.B \-L
Log energy computation flag.
If this option is specified, any field element
specified with
.B \-E
is assumed to be an energy, and its natural logarithm is computed
to provide a log energy term for the output.
If
.B \-L
is not specified, the logarithm is not computed, and the unmodified
field value is used.
This option has no effect if
.B \-E
is not specified.
.TP
.BI \-O " field"
.TP
\fB\-O "\fIfield\fB\^[\^\fIindex\fB\^]\^"\fR
A field element containing a ``zeroth-order cepstral parameter''
to be included in the output in addition to, or in place of,
the log energy term specified with
.BR \-E .
See section 5.5 (``Energy Measures'') of ref. [1].
This option is valid only when the basic parameter
kind specified with
.B \-k
is MFCC.
The resulting output parameter kind is MFCC_O if
.B \-O
is specified alone, and MFCC_O_E if both
.B \-O
and
.B \-E
are specified.
In the second form of the option argument, the
.I index
in brackets is an integer that indicates the field element to be used,
(counting from 0 for the first element).
(The reason for the double quotes is simply to prevent the shell
from giving the usual special interpretation to the brackets.)
The first form, without brackets, may be used
when the field has only one element.
.TP
.BI \-P " param_file" "\h'|1.4i'\fR(params)\fP"
The name of the ESPS parameter file.
.SH ESPS PARAMETERS
.PP
The terms ``parameter'' and ``parameter file'' are used in two senses.
An HTK parameter file, as described in ref. [1],
is a data file containing sample vectors
of parameters such as reflection coefficients, cepstral coefficients,
or filter-bank channel outputs.
An ESPS parameter file, as described in ref. [2],
is an ASCII file that contains values for parameters,
such as those described in this section,
that control the operation of a program.
.PP
The following parameters are read from the ESPS parameter file, if present.
.TP
.I "compute_log \- string"
.IP
A value of ``YES'' or ``yes'' means compute the logarithm of the field,
if any, specified with the option
.B \-E
or the parameter
.I Efield.
A value of ``NO'' or ``no'' means skip the logarithm computation.
This parameter is not read if the option
.B \-L
is specified.
The default value is ``NO''.
.TP
.I "field \- string"
.IP
The name of the field containing the waveform sample values
or parameter values to be included in the output.
As in the argument of the option
.BR \-f ,
the field name may be followed by a bracketed
.I grange
specification to indicate a subset of the field elements
that are to be included in the output.
See
.B \-f
under OPTIONS, and see the
.IR grange_switch (3\-ESPS)
man page.
This parameter is not read if
.B \-f
is specified.
It is required if
.B \-f
is not specified.
(There is no default.)
.TP
.I "nan \- int"
.IP
The total number of input records to process.
A value of 0, the default,
means continue processing until the end of the file is reached.
This parameter is not read if the option
.B \-r
is specified.
.TP
.I "parmKind \- string"
.IP
The basic kind of data to be indicated in the output HTK header.
Allowed values are those listed for the option
.BR \-k .
This parameter is not read if
that option is specified.
It is required if the option is not specified.

ESPS/general/src/featohtk/featohtk.c
/*
 * This material contains unpublished, proprietary software of
 * Entropic Research Laboratory, Inc. Any reproduction, distribution,
 * or publication of this work must be authorized in writing by Entropic
 * Research Laboratory, Inc., and must bear the notice:
 *
 *    "Copyright (c) 1998 Entropic Research Laboratory, Inc.
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended
 * publication of this source code.
 *
 * Written by:  Rod Johnson
 * Checked by:
 * Revised by:
 *
 * Brief description:
 *   HTK input filter for ESPS FEA files.
 */

static char *sccs_id = "@(#)featohtk.c	1.1\t3/17/98\tERL";

/* INCLUDE FILES */

#include <stdlib.h>
#include <math.h>
#include <esps/esps.h>
#include <esps/fea.h>
#include <esps/feasd.h>
#include <esps/array.h>
#include <esps/types.h>
#include <esps/range_switch.h>
#include <esps/strlist.h>

/* LOCAL CONSTANTS */

#define EC_SCCS_DATE	"3/17/98"
#define EC_SCCS_VERSION	"1.1"
#define MAX_INT		(0x7fffffff)	/* INT_MAX for EDR machines */
#define MAX_SHORT	(0x7fff)	/* SHRT_MAX for EDR machines */
#define SD_BUF_LEN	(1000)		/* length of FEA_SD intput buffer */

/* LOCAL MACROS */

#define ERROR(text) { \
  (void) fprintf(stderr, "%s: %s - exiting\n", ProgName, (text)); \
  exit(1);}

#define REQUIRE(test, text) {if (!(test)) ERROR(text)}

#define SYNTAX \
USAGE("featohtk [-f field[range]] [-k parmKind] [-r range] [-x debug_level]\n\t[-E field[index]] [-L] [-O field[index]] [-P param_file] input output")

/* LOCAL TYPEDEFS AND STRUCTURES */

/* HTK DEFINITIONS AND DECLARATIONS */
    
    /* These are taken from the HTK version 2.1.1 source files
     * HTKLib/HParm.h and HTKLib/HParm.c.
     * They may need to be revised when a later version is released.
     * Simply including HParm.h (the "right thing to do") required
     * including a cascade of further header files, extending to the
     * Esignal header file esignal.h, which contains definitions that
     * conflict with definitions in esps/esps.h.
     * Likewise linking the HTK libraries and the ESPS libraries in one
     * binary resulted in multiple definitions of numerous symbols
     * (apparently in the license manager).
     * All in all, too much complication to be worth doing.
     */

enum
{
    WAVEFORM,   LPC,        LPREFC,     LPCEPSTRA,
    LPDELCEP,   IREFC,      MFCC,       FBANK,
    MELSPEC,    USER,       DISCRETE,   ANON
};

static char	*pmkmap[] =
{
    "WAVEFORM", "LPC",      "LPREFC",   "LPCEPSTRA",
    "LPDELCEP", "IREFC",    "MFCC",     "FBANK",
    "MELSPEC",  "USER",     "DISCRETE", "ANON",
    NULL /* NULL termination, not in HTK source, needed for lin_search2 */
};

typedef short	ParmKind;

#define HASENERGY   (0100)
#define HASZEROC  (020000)
#define BASEMASK     (077)

/* SYSTEM FUNCTIONS AND VARIABLES */

extern int	getopt();
extern int	optind;
extern char	*optarg;

/* ESPS FUNCTIONS AND VARIABLES */

int		debug_level = 0;

/* LOCAL FUNCTION DECLARATIONS */

/* STATIC (LOCAL) GLOBAL VARIABLES */
 
static char	*ProgName = "featohtk";
static char	*Version = EC_SCCS_VERSION;
static char	*Date = EC_SCCS_DATE;

static char	*no_yes[] = {"NO", "YES", NULL};

/* MAIN PROGRAM */

int
main(argc, argv)
    int 	    argc;
    char	    **argv;
{
    int	    	    ch;			/* command-line option letter */

    char	    *fld = NULL;	/* basic param or sample field
					 * name with range */
    char	    *fld_name;		/* field name part of fld */
    long	    *fld_range;		/* indices of subrange of fld */
    long	    rng_len;		/* length of subrange of fld */
    int		    contig;		/* is fld_range a single sequence
					 * of consecutive integers without
					 * gaps? */
    int 	    fld_type;		/* data type of field named fld */
    char	    *fld_data;		/* data from subrange of fld */
    char	    *fldptr;		/* pointer to field data in rec */
    long	    fld_length;		/* length of field named fld */
    char	    *kind = NULL;	/* parameter kind string */
    short	    parmKind;		/* parameter kind binary code */
    int		    out_type;		/* output data type */
    char	    *rrange = NULL;	/* range of records to process */
    long	    start_rec;		/* number of first record to process */
    long	    end_rec;		/* number of last record to process */
    long	    num_recs;		/* number of records to process */
    int		    nSamples;		/* num_recs as int */
    char	    *Efld = NULL;	/* energy field name with index */
    char	    *Efld_name;		/* field name part of Efld */
    long	    *Efld_range;	/* pointer to index in Efld */
    long	    Erng_len;		/* length of subrange in Efld */
    int 	    Efld_type;		/* data type of E field (energy) */
    char	    *Eptr = NULL;	/* pointer to E data in rec */
    int 	    log_requested = NO;	/* compute log of energy? */
    char	    *comp_log_param;	/* value of "compute_log" in ESPS
					 * param file */
    char	    *Ofld = NULL;	/* 0-th cepstral coeff field name */
    char	    *Ofld_name;		/* field name part of Ofld */
    long	    *Ofld_range;	/* pointer to index in Ofld */
    long	    Orng_len;		/* length of subrange in Ofld */
    int 	    Ofld_type;		/* data type of O field (0-th order
					 * cepstral coefficient) */
    char	    *Optr = NULL;	/* pointer to O data in rec */
    char    	    *param_name = NULL;	/* parameter file name */
    char	    *inname;		/* input file name */
    FILE	    *infile;		/* input stream */
    struct header   *hd;		/* input file header */
    double	    record_freq;	/* input file record frequency (Hz) */
    int		    sampPeriod;		/* sampling period (100 ns) */
    double	    sampP_dbl;		/* sampPeriod as double */
    long	    ndrec;		/* input header item: number of
					 * data records */
    int		    fea_sd;		/* is input file FEA_SD? */
    struct fea_data *rec;		/* FEA input data record */
    struct feasd    *sd_rec;		/* FEA_SD input data structure */
    long	    chunk_len;		/* number of FEA_SD samples to try
					 * to read at one time */
    long	    act_len;		/* number of samples actually read */
    int		    eof;		/* end of input file reached? */
    char	    *outname;		/* output file name */
    FILE	    *outfile;		/* output stream */
    long	    samp_len;		/* number of items in output sample */
    int		    typesize;		/* size of one output item */
    short	    sampSize;		/* number of bytes in output sample */
    int		    out_len;		/* number of items of output sample
					 * actually written */
    char	    *buf;		/* output data buffer */
    float	    *Ebuf;		/* pointer to E data in buf */
    float	    *Obuf;		/* pointer to O data in buf */
    long	    num_read;		/* number of input records read
					 * or skipped over */
    long	    i, j;		/* loop and array indices */

    /*
     * PARSE COMMAND-LINE OPTIONS.
     */

    while ((ch = getopt(argc, argv, "f:k:r:x:E:LO:P:")) != EOF)
	switch (ch)
	{
	case 'f':
	    fld = optarg;
	    break;
	case 'k':
	    kind = optarg;
	    break;
	case 'r':
	    rrange = optarg;
	    break;
	case 'x':
	    debug_level = atoi(optarg);
	    break;
	case 'E':
	    Efld = optarg;
	    break;
	case 'L':
	    log_requested = YES;
	    break;
	case 'O':
	    Ofld = optarg;
	    break;
	case 'P':
	    param_name = optarg;
	    break;
	default:
	    SYNTAX;
	    break;
	}

    /*
     * PROCESS FILE NAMES AND OPEN FILES.
     */

    if (argc - optind > 2)
    {
	Fprintf(stderr,
		"%s: too many file names specified.\n", ProgName);
	SYNTAX;
    }
    if (argc - optind < 2)
    {
	Fprintf(stderr,
		"%s: too few file names specified.\n", ProgName);
	SYNTAX;
    }

    inname = eopen(ProgName,
		   argv[optind++], "r", FT_FEA, NONE, &hd, &infile);

    if (debug_level)
	Fprintf(stderr, "%s: input file \"%s\"\n", ProgName, inname);

    if (infile != stdin)
	REQUIRE(strcmp(inname, argv[optind]),
		"output file same as input");

    outname = eopen(ProgName,
		    argv[optind++], "w", NONE, NONE, NULL, &outfile);

    if (debug_level)
	Fprintf(stderr, "%s: output file: \"%s\"\n", ProgName, outname);

    /*
     * GET PARAMETER VALUES.
     */

    (void) read_params(param_name, SC_NOCOMMON, inname);

    /* Parameter kind (-k) */

    if (kind == NULL && symtype("parmKind") == ST_STRING)
    {
	kind = getsym_s("parmKind");
    }

    REQUIRE (kind != NULL,
	     "can't determine parameter kind");

    if (debug_level)
	Fprintf(stderr, "%s: parameter kind \"%s\"\n", ProgName, kind);

    parmKind = (short) lin_search2(pmkmap, kind);

    if (debug_level)
	Fprintf(stderr, "%s: parmKind code %x\n", ProgName, parmKind);

    switch (parmKind)
    {
	/* Most cases FALL THROUGH. */
    case WAVEFORM:
    case IREFC:
    case DISCRETE:
	out_type = SHORT;
	typesize = 2;
	break;
    case LPC:
    case LPREFC:
    case LPCEPSTRA:
    case MFCC:
    case FBANK:
    case MELSPEC:
    case USER:
	out_type = FLOAT;
	typesize = 4;
	break;
    case LPDELCEP:
	ERROR("delta-parameter fields not supported");
	break;
    default:
	ERROR("unsupported parmKind");
	break;
    }

    if (debug_level)
	Fprintf(stderr, "%s: out_type %d (%s), typesize %d\n",
		ProgName, out_type, ((out_type == SHORT) ? "SHORT"
				     : (out_type == FLOAT) ? "FLOAT"
				     : "other"),
		typesize);

    /* Base parameter field (-f) */

    if (fld == NULL && symtype("field") == ST_STRING)
    {
	fld = getsym_s("field");
    }

    REQUIRE (fld != NULL,
	     "can't determine field name");

    if (debug_level)
	Fprintf(stderr, "%s: base parameter field \"%s\"\n",
		ProgName, fld);

    fld_range = fld_range_switch(fld, &fld_name, &rng_len, hd);
    fld_type = get_fea_type(fld_name, hd);
    fld_length = get_fea_siz(fld_name, hd, NULL, NULL);

    if (debug_level)
    {
	Fprintf(stderr, "%s: field name \"%s\", range length %ld\n",
		ProgName, fld_name, rng_len);
	Fprintf(stderr, "%s: field type %d (%s), field length %ld\n",
		ProgName, fld_type, type_codes[fld_type], fld_length);
    }

    REQUIRE(is_type_numeric(fld_type),
	    "field undefined or non-numeric in input file header");
    spsassert(fld_range != NULL,
	      "fld_range_switch returned NULL");

    if (debug_level >= 2)
    {
	Fprintf(stderr, "%s: field elements\n\t[", ProgName);
	for (i = 0; i < rng_len; i++)
	    Fprintf(stderr, "%ld%s", fld_range[i],
		    (i == rng_len - 1) ? "]\n"
		    : (i%10 == 9) ? ",\n\t "
		    : ", ");
    }

    if (parmKind == WAVEFORM)
	REQUIRE(rng_len == 1,
		"range size must be 1 for WAVEFORM");

    contig = YES;
    for (i = 0; i < rng_len; i++)
    {
	j = fld_range[i];
	REQUIRE(j >= 0 && j < fld_length,
		"index out of range for field");
	if (j != fld_range[0] + i)
	    contig = NO;
    }

    if (debug_level)
	Fprintf(stderr, "%s: range %s contiguous\n",
		ProgName, (contig) ? "IS" : "is NOT");

    /* Energy parameter field (-E) */

    if (Efld == NULL && symtype("Efield") == ST_STRING)
    {
	Efld = getsym_s("Efield");
    }

    if (debug_level)
	Fprintf(stderr, "%s: E field \"%s\"\n",
		ProgName, (Efld == NULL) ? "<null>" : Efld);

    if (Efld != NULL)
    {
	REQUIRE(out_type == FLOAT,
		"energy term not supported with non-FLOAT output");
	Efld_range = fld_range_switch(Efld, &Efld_name, &Erng_len, hd);
	Efld_type = get_fea_type(Efld_name, hd);

	if (debug_level)
	{
	    Fprintf(stderr, "%s: Efld name \"%s\", range len %ld\n",
		    ProgName, Efld_name, Erng_len);
	    Fprintf(stderr, "%s: Efld type %d (%s)\n",
		    ProgName, Efld_type, type_codes[Efld_type]);
	    Fprintf(stderr, "%s: Efld element %ld\n", ProgName, Efld_range[0]);
	}

	REQUIRE(is_type_numeric(Efld_type),
		"Energy field undefined or non-numeric in input file header");
	spsassert(Efld_range != NULL,
		  "fld_range_switch returned NULL");
	REQUIRE(Erng_len == 1,
		"Energy field must have exactly one element");
	REQUIRE(Efld_range[0] >= 0
		&& Efld_range[0] < get_fea_siz(Efld_name, hd, NULL, NULL),
		"index out of range for Energy field");

	parmKind |= HASENERGY;

	if (debug_level)
	    Fprintf(stderr, "%s: now parmKind code is %x\n",
		    ProgName, parmKind);
    }

    /* Log energy computation flag (-L) */

    if (!log_requested && symtype("compute_log") == ST_STRING)
    {
	comp_log_param = getsym_s("compute_log");

	switch (lin_search(no_yes, comp_log_param))
	{
	case 0:
	    /* value remains NO */
	    break;
	case 1:
	    log_requested = YES;
	    break;
	default:
	    ERROR("ambiguous value of parameter \"compute_log\"");
	    break;
	}
    }

    if (debug_level)
	Fprintf(stderr, "%s: log computation %s requested\n",
		ProgName, (log_requested) ? "IS" : "is NOT");

    /* Zeroth cepstral parameter field (-O) */

    if (Ofld == NULL && symtype("Ofield") == ST_STRING)
    {
	Ofld = getsym_s("Ofield");
    }

    if (debug_level)
	Fprintf(stderr, "%s: O field \"%s\"\n",
		ProgName, (Ofld == NULL) ? "<null>" : Ofld);

    if (Ofld != NULL)
    {
	REQUIRE((parmKind & BASEMASK) == MFCC,
		"for 0-order cepstral coefficient, paramKind must be MFCC");
	Ofld_range = fld_range_switch(Ofld, &Ofld_name, &Orng_len, hd);
	Ofld_type = get_fea_type(Ofld_name, hd);

	if (debug_level)
	{
	    Fprintf(stderr, "%s: Ofld name \"%s\", range len %ld\n",
		    ProgName, Ofld_name, Orng_len);
	    Fprintf(stderr, "%s: Ofld type %d (%s)\n",
		    ProgName, Ofld_type, type_codes[Ofld_type]);
	    Fprintf(stderr, "%s: Ofld element %ld\n", ProgName, Ofld_range[0]);
	}

	REQUIRE(is_type_numeric(Ofld_type),
		"Zero-order cepstral field undefined or non-numeric in input file header");
	spsassert(Ofld_range != NULL,
		  "fld_range_switch returned NULL");
	REQUIRE(Orng_len == 1,
		"Zero-order cepstral field must have exactly one element");
	REQUIRE(Ofld_range[0] >= 0
		&& Ofld_range[0] < get_fea_siz(Ofld_name, hd, NULL, NULL),
		"index out of range for 0-order cepstral coefficient field");

	parmKind |= HASZEROC;

	if (debug_level)
	    Fprintf(stderr, "%s: now parmKind code is %x\n",
		    ProgName, parmKind);
    }

    /* Range of records to process (-r) */

    if (rrange != NULL)
    {
	start_rec = 1;
	end_rec = LONG_MAX;
	lrange_switch(rrange, &start_rec, &end_rec, 0);
	num_recs = (end_rec != LONG_MAX)
	    ? end_rec - (start_rec - 1)
		: 0;
    }
    else
    {
	start_rec =
	    (symtype("start") == ST_INT)
		? getsym_i("start")
		    : 1;
	num_recs =
	    (symtype("nan") == ST_INT)
		? getsym_i("nan")
		    : 0;
	end_rec =
	    (num_recs != 0)
		? start_rec - 1 + num_recs
		    : LONG_MAX;
    }

    if (debug_level)
	Fprintf(stderr, "%s: start_rec %ld, end_rec %ld, num_recs %ld\n",
		ProgName, start_rec, end_rec, num_recs);
	

    if (num_recs == 0)			/* implies continue to end of file */
    {
	ndrec = hd->common.ndrec;
	REQUIRE(ndrec != -1,		/* pipe or variable record length */
		"explicit end of range needed for piped input");

	end_rec = ndrec;
	num_recs = end_rec - (start_rec - 1);

	if (debug_level)
	    Fprintf(stderr, "%s: now end_rec %ld, num_recs %ld\n",
		ProgName, end_rec, num_recs);
    }

    REQUIRE(start_rec >= 1,
	    "can't start before beginning of file");
    REQUIRE(end_rec >= start_rec,
	    "empty range of records specified");

    /*
     * SET UP I/O BUFFERS.
     */

    fea_sd = (parmKind == WAVEFORM
	      && hd->hd.fea->fea_type == FEA_SD
	      && strcmp(fld_name, "samples") == 0
	      && fld_length == 1);

    if (fea_sd)  /* prepare to use FEA_SD support functions for input */
    {
	if (debug_level)
	    Fprintf(stderr, "%s: allocating FEA_SD record structure\n",
		    ProgName);

	sd_rec = allo_feasd_recs(hd, out_type, SD_BUF_LEN, NULL, NO);
	REQUIRE(sd_rec != NULL,
		"FEA_SD record allocation failed");
	buf = sd_rec->data;

	samp_len = 1;

	if (debug_level)
	    Fprintf(stderr, "%s: samp_len %ld item\n", ProgName, samp_len);
    }
    else	 /* prepare to use general FEA support functions for input */
    {
	if (debug_level)
	    Fprintf(stderr, "%s: allocating general FEA record\n",
		    ProgName);

	rec = allo_fea_rec(hd);

	REQUIRE(rec != NULL,
		"FEA record allocation failed");

	fldptr = get_fea_ptr(rec, fld_name, hd);

	if (Efld != NULL)
	    Eptr = get_fea_ptr(rec, Efld_name, hd);

	if (Ofld != NULL)
	    Optr = get_fea_ptr(rec, Ofld_name, hd);

	samp_len = rng_len;
	if (Ofld != NULL)
	    samp_len++;
	if (Efld != NULL)
	    samp_len++;

	if (debug_level)
	    Fprintf(stderr, "%s: samp_len %ld items\n", ProgName, samp_len);

	buf = ((fld_type == out_type && contig && Ofld == NULL && Efld == NULL)
	       ? fldptr + typesize * fld_range[0]
	       : arr_alloc(1, &samp_len, out_type, 0));

	fld_data = ((contig) ? fldptr + typesiz(fld_type) * fld_range[0]
		    : (fld_type == out_type) ? buf
		    : arr_alloc(1, &rng_len, fld_type, 0));

	if (debug_level)
	    Fprintf(stderr, "%s: field ptr %p, fld_data %p, buf %p\n",
		    ProgName, fldptr, fld_data, buf);

	if (Ofld != NULL)
	{
	    Obuf = (float *) buf + rng_len;

	    if (debug_level)
		Fprintf(stderr, "%s: Obuf %p\n", ProgName, Obuf);
	}

	if (Efld != NULL)
	{
	    Ebuf = (Ofld != NULL) ? Obuf + 1 : (float *) buf + rng_len;

	    if (debug_level)
		Fprintf(stderr, "%s: Ebuf %p\n", ProgName, Ebuf);
	}
    }

    /*
     * WRITE OUTPUT HEADER INFO.
     */

    REQUIRE(num_recs <= MAX_INT,
	    "range of records too long");
    nSamples = (int) num_recs;
    miio_put_int(&nSamples, 1, YES, outfile);

    record_freq = get_genhd_val("record_freq", hd, 0.0);
    REQUIRE(record_freq > 0.0,
	    "header item \"record_freq\" missing or invalid");
    sampP_dbl = floor(0.5 + 1.0e7/record_freq);
    REQUIRE(sampP_dbl >= 1 && sampP_dbl <= MAX_INT,
	    "header item \"record_freq\" invalid");
    sampPeriod = (int) sampP_dbl;
    miio_put_int(&sampPeriod, 1, YES, outfile);

    REQUIRE(samp_len <= MAX_SHORT/typesize,
	    "parameter field too big");
    sampSize = samp_len * typesize;
    miio_put_short(&sampSize, 1, YES, outfile);

    miio_put_short(&parmKind, 1, YES, outfile);

    if (debug_level)
    {
	Fprintf(stderr,
		"%s: nSamples %d, sampPeriod %d, sampSize %d, parmKind %d\n",
		ProgName, nSamples, sampPeriod, sampSize, parmKind);
    }

    /*
     * MAIN READ-WRITE LOOP.
     */

    num_read = start_rec - 1;

    if (debug_level)
	Fprintf(stderr, "%s: skipping %ld initial records\n",
		ProgName, num_read);

    fea_skiprec(infile, num_read, hd);

    if (fea_sd)  /* use FEA_SD support functions for input */
    {
	chunk_len = MIN(end_rec - num_read, SD_BUF_LEN);
	eof = NO;

	while (chunk_len > 0 && !eof)
	{
	    if (debug_level >= 2)
		Fprintf(stderr, "%s: reading chunk of %ld samples . . . .\n",
			ProgName, chunk_len);

	    act_len = get_feasd_recs(sd_rec, 0L, chunk_len, hd, infile);

	    if (debug_level >= 2)
		Fprintf(stderr, "%s: got %ld samples\n",
			ProgName, act_len);

	    num_read += act_len;

	    if (debug_level >= 2)
		Fprintf(stderr, "%s: read sample %ld\n",
			ProgName, num_read);

	    out_len = miio_put_short((short *) buf,
				     (int) act_len, YES, outfile);

	    if (debug_level >= 2)
		Fprintf(stderr, "%s: wrote %d samples\n", ProgName, out_len);

	    if (out_len != act_len)
	    {
		if (debug_level >= 2)
		    Fprintf(stderr,
			    "%s: failed to write FEA_SD output chunk - exiting\n",
			    ProgName);

		exit(1);
	    }
	    
	    eof = act_len < chunk_len;
	    if (end_rec - num_read < SD_BUF_LEN)
		chunk_len = end_rec - num_read;
	}
    }
    else         /* use general FEA support functions for input */
    {
	while (num_read < end_rec && get_fea_rec(rec, hd, infile) != EOF)
	{
	    num_read++;

	    if (debug_level >= 2)
		Fprintf(stderr, "%s: read record %ld . . . .\n",
			ProgName, num_read);

#define COPY(TYPE) \
	    {   \
		TYPE    *src = (TYPE *) fldptr; \
		TYPE    *dst = (TYPE *) fld_data; \
		for (i = 0; i < rng_len; i++) dst[i] = src[fld_range[i]]; \
	    }

	    if (!contig)
	    {
		if (debug_level >= 2)
		    Fprintf(stderr, "%s: copying subrange\n", ProgName);

		switch (fld_type)
		{
		case DOUBLE:      COPY(double);      break;
		case FLOAT:       COPY(float);       break;
		case LONG:        COPY(long);        break;
		case SHORT:       COPY(short);       break;
		case BYTE:        COPY(char);        break;
		case DOUBLE_CPLX: COPY(double_cplx); break;
		case FLOAT_CPLX:  COPY(float_cplx);  break;
		case LONG_CPLX:   COPY(long_cplx);   break;
		case SHORT_CPLX:  COPY(short_cplx);  break;
		case BYTE_CPLX:   COPY(byte_cplx);   break;
		}
	    }

	    if (buf != fld_data)
	    {
		if (debug_level >= 2)
		    Fprintf(stderr, "%s: type-converting subrange\n",
			    ProgName);

		(void) type_convert(rng_len, fld_data, fld_type,
				    buf, out_type, warn_on_clip);
	    }

	    if (Ofld != NULL)
	    {
		if (debug_level >= 2)
		    Fprintf(stderr, "%s: copying or converting Ofld data\n",
			    ProgName);

		if (Ofld_type == FLOAT)
		    *Obuf = *(float *) Optr;
		else
		    (void) type_convert(1L, (char *) Optr, Ofld_type,
					(char *) Obuf, FLOAT, warn_on_clip);
	    }

	    if (Efld != NULL)
	    {
		if (debug_level >= 2)
		    Fprintf(stderr, "%s: copying or converting Efld data\n",
			    ProgName);

		if (Efld_type == FLOAT)
		    *Ebuf = *(float *) Eptr;
		else
		    (void) type_convert(1L, (char *) Eptr, Efld_type,
					(char *) Ebuf, FLOAT, warn_on_clip);
		if (log_requested)
		{
		    if (debug_level >= 2)
			Fprintf(stderr, "%s: computing log of Efld data\n",
				ProgName);

		    *Ebuf = log(*Ebuf);
		}
	    }

	    switch (out_type)
	    {
	    case SHORT:
		out_len = miio_put_short((short *) buf,
					 (int) samp_len, YES, outfile);
		break;
	    case FLOAT:
		out_len = miio_put_float((float *) buf,
					 (int) samp_len, YES, outfile);
		break;
	    default:
		spsassert(0, "invalid data type for output");
	    }

	    if (debug_level >= 2)
		Fprintf(stderr, "%s: wrote %d items\n", ProgName, out_len);

	    if (out_len != samp_len)
	    {
		if (debug_level)
		    Fprintf(stderr,
			    "%s: failed to write output vector - exiting\n",
			    ProgName);
		exit(1);
	    }
	}
    }

    if (debug_level)
	Fprintf(stderr, "%s: end of loop; read %ld records\n",
		ProgName, num_read);
    
    REQUIRE(num_read == end_rec,
	    "premature end of file");

    exit(0);
    /*NOTREACHED*/
}
